pfzip.1
.Dd Feb 28, 2025
.Dt PFZIP 1
.Os
.Sh NAME
.Nm pfzip
.Nd archive physical files and streamfiles
.Sh SYNOPSYS
.Nm
.Op Fl EprstWV
.Ar zip-file
.Ar files
.Sh DESCRIPTION
The
.Nm
utility creates a Zip format archive specified in the
.Ar zip-file
argument, containing the physical file members or IFS streamfiles as specified in the
.Ar files
argument. The files inside will be converted to the PASE locale, and for physical
file members, have Unix line endings. The file extension of a physical file
 member will also be changed to match its source type, for example, "RPGLE". The
Zip archive will also have comments for each file that match the physical file
members' text descriptions.
.Pp
The options are as follows:
.Bl -tag -width indent
.It Fl E
Don't translate the path of physical file members in the archive.
.It Fl p
Searches non-source physical files. Note that non-source physical files are
subject to
.Lk https://www.ibm.com/docs/en/i/7.5?topic=qsyslib-file-handling-restrictions-in-file-system some limitations
as they are read in POSIX binary mode.
.It Fl r
Recurses into IFS directories, libraries, and physical files.
.It Fl s
Don't print error messages; the return code is unchanged.
.It Fl t
Don't trim whitespace at the end of lines; by default, whitespace is trimmed.
This preserves the padding to match the length of the record.
.It Fl W
Overwrite the archive if it exists already.
.It Fl V
Print the version number of the utility and any libraries it uses.
.El
.Sh EXAMPLES
Put the library QSYSINC into a zip file called includes.zip:
.Pp
.Dl pfzip -r includes.zip /QSYS.LIB/QSYSINC.LIB
.Pp
Note that the
.Fl r
flag like for pfgrep will make it recurse into directories.
.Pp
The member descriptions, record length, original CCSID before translation, and
the source type applied to the file extension (if you don't want the new file
extension, pass the
.Fl E
flag) will be in the archive. You can view it in an archiver utility like
.Xr unzip 1
and see the descriptions like so:
.Bd -literal
Archive:  includes.zip
  Length      Date    Time    Name
---------  ---------- -----   ----
     9089  2013-10-06 12:59   QSYS.LIB/QSYSINC.LIB/ARPA.FILE/INET.C
(original PF record length 80 CCSID 37)
    38324  2013-10-06 13:00   QSYS.LIB/QSYSINC.LIB/ARPA.FILE/NAMESER.C
(original PF record length 80 CCSID 37)
     6790  2013-10-06 12:59   QSYS.LIB/QSYSINC.LIB/ARPA.FILE/REXEC.C
REMOTE EXECUTION APIS                              (original PF record length 80 CCSID 37)
[...]
.Ed
.Pp
You can also compose pfzip with pfgrep together. For example, you can put all
files that match a regular expression like so:
.Pp
.Dl pfzip client.zip $(pfgrep -l Qo (sales|exports) Qc /QSYS.LIB/PRODLIB.LIB/Q*SRC.LIB)
.Pp
To break this down, the
.Ql $()
shell construct takes the output of one command and turns it into command line
arguments for another command. The
.Fl l
flag to pfgrep will make it only print the files that match instead of the
matching text in the files.
.Sh SEE ALSO
.Xr pfcat 1 ,
.Xr pfgrep 1 ,
.Xr pfstat 1 ,
.Xr libzip 3 ,
.Xr unzip 1
.Sh AUTHORS
The
.Nm
utility was written for Seiden Group by
.An Calvin Buckley Aq Mt calvin@seidengroup.com
and
.Lk https://github.com/SeidenGroup/pfgrep/graphs/contributors other contributors .


pfzip.c
/*
 * Copyright (c) 2024-2025 Seiden Group
 *
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

#include <as400_protos.h>
#include <as400_types.h>
#include <dirent.h>
#include <fcntl.h>
#include <limits.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <sys/errno.h>
#include <sys/mode.h>
#include <sys/stat.h>
#include <unistd.h>

#include <zip.h>

#include "common.h"
#include "errc.h"

static void usage(char *argv0)
{
	fprintf(stderr, "usage: %s [-EprstWV] output_file.zip files\n", argv0);
}

static char *ends_with_mbr(const char *str)
{
	char *ext = strrchr(str, '.');
	if (ext && strcasecmp("MBR", ext + 1) == 0) {
		return ext + 1;
	}
	return NULL;
}

static bool is_nul_or_space(const char c)
{
	return c == ' ' || c == '\0';
}

/**
 * Change the filename to be better suited to an archive. This can remove
 * the leading / from a path to make it not absolute, and replace the file
 * extension of a PF member to match its source type when possible.
 */
static void normalize_path(char *output, size_t output_size, File *file, bool replace_mbr_ext)
{
	// If requested, we can replace the generic .MBR suffix on a member
	// with a file extension derived from the member's source type.
	// XXX: Condense these checks, make case insensitive
	if (replace_mbr_ext && !is_nul_or_space(*file->source_type)
		&& strstr(file->filename, "/QSYS.LIB/") != NULL
		&& ends_with_mbr(file->filename) != NULL) {
		char new_path[PATH_MAX + 1];
		// +1 to eliminate leading / for abs path
		strncpy(new_path, file->filename + 1, sizeof(new_path));
		// Copy without whitespace
		char *extension = ends_with_mbr(new_path);
		const char *source_type = file->source_type;
		while (!is_nul_or_space(*source_type)) {
			*extension++ = *source_type++;
		}
		*extension = '\0';
		strncpy(output, new_path, output_size);
		return;
	}
	// Otherwise, just use the path as-is, removing absoluteness.
	const char *input = file->filename;
	if (input[0] == '/') {
		input++;
	}
	strncpy(output, input, output_size);
}

int do_action(pfgrep *state, File *file)
{
	int ret = 1, nonfatal_ret;
	const char *buf = state->conv_buffer;
	if (file->record_length == 0 && file->ccsid == state->pase_ccsid) {
		buf = state->read_buffer;
	}
	size_t len = strlen(buf);
	// we must keep a copy around until zip_close, and we reread the buffer
	// therefore make a copy (NBD) and tell libzip to free (last parm)
	char *buf_copy = strdup(buf);
	zip_source_t *s = zip_source_buffer(state->archive, buf_copy, len, 1);
	if (s == NULL && !state->silent) {
		fprintf(stderr, "zip_source_buffer(%s): %s\n",
			file->filename,
			zip_strerror(state->archive));
		ret = -1;
		goto fail;
	}

	char path[PATH_MAX + 1];
	// We'll need the member metadata for the comment and extension.
	if (file->record_length != 0) {
		get_member_info(file);
	}
	normalize_path(path, sizeof(path), file, !state->dont_replace_extension);
	zip_int64_t index = zip_file_add(state->archive, path, s, 0);
	if (index == -1 && !state->silent) {
		fprintf(stderr, "zip_file_add(%s): %s\n",
			file->filename,
			zip_strerror(state->archive));
		zip_source_free(s);
		ret = -1;
		goto fail;
	}

	char comment[512];
	// Put the member description as a comment.
	// The other metdata is there too; may not be best place for it
	if (file->record_length == 0) {
		snprintf(comment, 512, "(original streamfile CCSID %d)", file->ccsid);
	} else if (*file->description) {
		snprintf(comment, 512, "%s (original PF record length %d CCSID %d)",
			file->description,
			file->record_length,
			file->ccsid);
	} else {
		snprintf(comment, 512, "(original PF record length %d CCSID %d)",
			file->record_length,
			file->ccsid);
	}
	// not critical if these fail, but do warn
	nonfatal_ret = zip_file_set_comment(state->archive, index, comment, strlen(comment), 0);
	if (nonfatal_ret && !state->silent) {
		fprintf(stderr, "zip_file_set_comment: Can't set comment for %s",
			file->filename);
	}
	nonfatal_ret = zip_file_set_mtime(state->archive, index, file->mtime, 0);
	if (nonfatal_ret && !state->silent) {
		fprintf(stderr, "zip_file_set_comment: Can't set modification time (%zd) for %s",
			file->mtime,
			file->filename);
	}
fail:
	return ret;
}

int main(int argc, char **argv)
{
	pfgrep state = {0};
	common_init(&state);

	int ch;
	while ((ch = getopt(argc, argv, "EprstWV")) != -1) {
		switch (ch) {
		case 'E':
			state.dont_replace_extension = true;
			break;
		case 'p':
			state.search_non_source_files = true;
			break;
		case 'r':
			state.recurse = true;
			break;
		case 's':
			state.silent = true;
			break;
		case 't':
			state.dont_trim_ending_whitespace = true;
			break;
		case 'W':
			state.overwrite = true;
			break;
		case 'V':
			print_version("pfzip");
			return 0;
		default:
			usage(argv[0]);
			return 3;
		}
	}

	if (optind + 1 >= argc) {
		usage(argv[0]);
		return 3;
	}
	const char *output_file = argv[optind++];
	state.file_count = argc - optind;
	if (state.file_count == 0 && !state.silent) {
		fprintf(stderr, "%s: need files for archive\n", argv[0]);
		return 5;
	}

	int zerrno;
	int open_flags = ZIP_CREATE;
	if (state.overwrite) {
		open_flags |= ZIP_TRUNCATE;
	}
	state.archive = zip_open(output_file, open_flags, &zerrno);
	if (state.archive == NULL && !state.silent) {
		zip_error_t error;
		zip_error_init_with_code(&error, zerrno);
		fprintf(stderr, "zip_open: %s\n",
        		zip_error_strerror(&error));
		zip_error_fini(&error);
		return 6;
	}

	bool any_match = false, any_error = false;
	for (int i = optind; i < argc; i++) {
		int ret = do_thing(&state, argv[i], false);
		if (ret > 0) {
			any_match = true;
		} else if (ret < 0) {
			any_error = true;
		}
	}

	if (zip_close(state.archive) == -1 && !state.silent) {
		fprintf(stderr, "zip_close: %s\n", zip_strerror(state.archive));
		return 4;
	}
#ifdef DEBUG
	// This deinitialization may be unnecessary, do it for future use of
	// sanitizers/*grind when available on i
	free_cached_iconv();
	free_cached_record_sizes();
	free(state.read_buffer);
	free(state.conv_buffer);
#endif

	return any_error ? 2 : (any_match ? 0 : 1);
}
